pilot/pkg/networking/plugin/authz/rbac_v2.go

```diff
@@ -0,0 +1,129 @@
+// generate any RBAC config by default.
+//
+// Changes from rbac_v2.go compared to rbac.go:
+// * Deprecate ServiceRoleBinding. Only support two CRDs: ServiceRole and AuthorizationPolicy.
+// * Allow multiple bindings and roles in one CRD, i.e. Authorization.
+// * Support workload selector.
+package authz
+
+import (
+	"fmt"
+
+	http_config "github.com/envoyproxy/go-control-plane/envoy/config/filter/http/rbac/v2"
+	policyproto "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v2alpha"
+
+	rbacproto "istio.io/api/rbac/v1alpha1"
+	"istio.io/istio/pilot/pkg/model"
+)
+
+// convertRbacRulesToFilterConfigV2 is the successor of convertRbacRulesToFilterConfig, which supports
+// converting AuthorizationPolicy.
+func convertRbacRulesToFilterConfigV2(service *serviceMetadata, option rbacOption) *http_config.RBAC {
+	rbac := &policyproto.RBAC{
+		Action:   policyproto.RBAC_ALLOW,
+		Policies: map[string]*policyproto.Policy{},
+	}
+
+	if option.globalPermissiveMode {
+		// Permissive Mode will be ignored.
+		rbacLog.Debugf("Permissive mode is not supported for service %s.", service.name)
+		// TODO(pitlv2109): Handle permissive mode in the future.
+		return &http_config.RBAC{Rules: rbac}
+	}
+
+	namespace := service.attributes[attrDestNamespace]
+	if option.authzPolicies == nil || option.authzPolicies.NamespaceToAuthorizationConfigV2 == nil {
+		return &http_config.RBAC{Rules: rbac}
+	}
+	if _, present := option.authzPolicies.NamespaceToAuthorizationConfigV2[namespace]; !present {
+		return &http_config.RBAC{Rules: rbac}
+	}
+	// Get all AuthorizationPolicy Istio config from this namespace.
+	allAuthzPolicies := option.authzPolicies.NamespaceToAuthorizationConfigV2[namespace].AuthzPolicies
+	for _, authzPolicy := range allAuthzPolicies {
+		// If a WorkloadSelector is used in the AuthorizationPolicy config and does not match this service, skip this
+		// AuthorizationPolicy.
+		serviceLabelsCollection := model.LabelsCollection{service.labels}
+		if authzPolicy.Policy.WorkloadSelector != nil &&
+			!(serviceLabelsCollection.IsSupersetOf(authzPolicy.Policy.WorkloadSelector.Labels)) {
+			continue
+		}
+		for i, binding := range authzPolicy.Policy.Allow {
+			// Check if we have converted this ServiceRole before.
+			permissions := make([]*policyproto.Permission, 0)
+			// Convert role.
+			role := option.authzPolicies.RoleForNameAndNamespace(namespace, binding.RoleRef.Name)
+			for _, rule := range role.Rules {
+				// Check to make sure the services field does not exist in the ServiceRole.
+				if len(rule.Services) > 0 {
+					rbacLog.Errorf("should not have services field in ServiceRole (found in %s)", binding.RoleRef.Name)
+					continue
+				}
+				// Check to make sure that the current service (caller) is the one this rule is applying to.
+				if !service.areConstraintsMatched(rule) {
+					rbacLog.Errorf("rule has constraints but doesn't match with the service (found in %s)", binding.RoleRef.Name)
```

Comment:
Should this be a "Info" log?

Reply:
Also, add a test case for it?

Reply:
I saw many "informative" logs in rbac.go are using `Errorf` level ([example](https://github.com/istio/istio/blob/f76c5f2e4b7d7dcbc587b741fa87d6fe0a8b2ce3/pilot/pkg/networking/plugin/authz/rbac.go#L506)), so I just followed the convention :D 

As for the test, this is simply a refactoring of the `match()` function in rbac.go. I just moved a block of code into a different function. There are tests for `match()` in rbac_test.go. Do you think we still need test for this sub-function?

Reply:
The example was using Debugf. In this case, it is just filtering out the applied roles based on "constraints", and is not an error condition. 

For test, we can have a test that has a role with "constraints" of labels (and will be filtered out here). What do you think?

Reply:
Oops, so do you think we should use Info or Debug? 

I can add a test if you think it's necessary, though I think [this](https://github.com/istio/istio/blob/authz-v2/pilot/pkg/networking/plugin/authz/rbac_test.go#L1880-L1907) should be good enough? The code for `areConstraintsMatched` used to be in `match()`, but I broke it into a sub-function.

Reply:
Let's use "Debug" here. 

The test case I had in mind is that, the role uses "constrians: key: destination.labels[app], values: ["a"]", The binding has selector "app: b". In this case, the role does not apply. It's basically testing combination of "selector" and "constraints" on labels.

Reply:
Ok so I created a new test case (and pushed). I have a role with destination labels app = foo, a service with label selector app = foo, and 2 AuthorizationPolicy, one with no workload selector, and one with selector app = bar. Only the first AuthorizationPolicy applies. So both constraints and selector work as expected in this test case.